File: src/sema/scope.ts
import { Symbol } from "./symbol";

export class Scope {
  private parent?: Scope;
  private symbols: Map<string, Symbol>;

  constructor(parent?: Scope) {
    this.parent = parent;
    this.symbols = new Map();
  }

  insert(name: string, symbol: Symbol): boolean {
    if (this.symbols.has(name)) {
      return false;
    }
    this.symbols.set(name, symbol);
    return true;
  }

  lookup(name: string): Symbol | undefined {
    const local = this.symbols.get(name);
    if (local) return local;
    if (this.parent) return this.parent.lookup(name);
    return undefined;
  }

  lookupLocal(name: string): Symbol | undefined {
    return this.symbols.get(name);
  }
}

File: src/sema/scope.nim
import std/tables
import ./symbol

type Scope* = ref object
  parent: Scope
  symbols: Table[string, Symbol]

proc newScope*(parent: Scope = nil): Scope = Scope(parent: parent, symbols: initTable[string, Symbol]())
proc insert*(scope: Scope; name: string; symbol: Symbol): bool =
  if name in scope.symbols: return false
  scope.symbols[name] = symbol
  true
proc lookup*(scope: Scope; name: string): tuple[found: bool, symbol: Symbol] =
  if name in scope.symbols: return (true, scope.symbols[name])
  if scope.parent != nil: return scope.parent.lookup(name)
  (false, Symbol())
proc lookupLocal*(scope: Scope; name: string): tuple[found: bool, symbol: Symbol] =
  if name in scope.symbols: (true, scope.symbols[name]) else: (false, Symbol())
